### ISSUE TITLE
opt: internal error: outer cols mismatch: (23) vs ()

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#36183]
I've seen this internal error a few times, but am unable to come up with SQL that recreates it from scratch. The exact offending query is below, but even after inserting some dummy rows into the table it doesn't produce this same error. I'm filing here so it's recorded. I'll continue trying to find an exact repro.

```
E190326 21:18:17.554665 4985 sql/sqltelemetry/report.go:39  [n1,client=127.0.0.1:58926,user=root] encountered internal error:
sql/opt/props/logical.go:481 in VerifyAgainst(): (XX000) internal error: outer cols mismatch: (20) vs ()
-- detail --
logical.go:481: outer cols mismatch: %s vs %s | (20); ()
github.com/cockroachdb/cockroach/pkg/sql/opt/props/logical.go:481: in VerifyAgainst()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:56: in checkExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:14695: in AddInnerJoinToGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:422: in exploreInnerJoin()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:22: in exploreGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182: in exploreGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:448: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:577: in enforceProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:475: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:209: in Optimize()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:429: in buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:154: in makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:987: in makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:874: in dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459: in execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101: in execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1189: in run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436: in ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:577: in func1()
E190326 21:18:17.560590 4985 sql/sqltelemetry/report.go:39  encountered internal error:
sql/opt/props/logical.go:481 in VerifyAgainst(): (XX000) internal error: outer cols mismatch: (20) vs ()
-- detail --
logical.go:481: outer cols mismatch: %s vs %s | (20); ()
github.com/cockroachdb/cockroach/pkg/sql/opt/props/logical.go:481: in VerifyAgainst()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/check_expr.go:56: in checkExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:14695: in AddInnerJoinToGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:422: in exploreInnerJoin()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:22: in exploreGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182: in exploreGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:448: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:577: in enforceProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:475: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:235: in optimizeExpr()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:490: in optimizeGroupMember()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:435: in optimizeGroup()
github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:209: in Optimize()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:429: in buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:154: in makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:987: in makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:874: in dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459: in execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101: in execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1189: in run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:436: in ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:577: in func1()
CRASHER:
caused by: server panic: pq: internal error: outer cols mismatch: (20) vs ()

STATEMENT:
INSERT
INTO
	defaultdb.public.table0 AS tab_7206
SELECT
	NULL AS col_21901,
	'09:02:27.316911':::TIME AS col_21902,
	tab_7207.col8 AS col_21903,
	(
		SELECT
			DISTINCT tab_7210.col3 AS col_21926
		FROM
			defaultdb.public.table0 AS tab_7208
			JOIN defaultdb.public.table0 AS tab_7209 ON tab_7207.col7
			FULL JOIN defaultdb.public.table0 AS tab_7210 ON false,
			[
				INSERT
				INTO
					defaultdb.public.table0 AS tab_7211
				WITH
					with_1508 (col_21904, col_21905, col_21906)
						AS (
							SELECT
								*
							FROM
								(
									VALUES
										(
											CASE
											WHEN (NULL OR (true AND ((((('181.241.110.49/1':::INET IN ()) OR (NOT CASE WHEN true THEN (true OR false) ELSE false END)) OR ((false OR true) < (NULL::INET << '8be9:206f:b99d:2705:d621:b555:3fb6:9083/76':::INET::INET))) OR true) OR true)))
											THEN array_position(NULL::INTERVAL[], '45 years 9 mons 523 days 15:32:34.568577':::INTERVAL::INTERVAL)::INT8
											ELSE (-8973059808895841774):::INT8
											END,
											'null':::JSONB,
											(-0.1462563335361856):::FLOAT8
										)
								)
									AS tab_7212 (col_21904, col_21905, col_21906)
						),
					with_1509 (col_21907, col_21908, col_21909, col_21910, col_21911, col_21912, col_21913)
						AS (
							SELECT
								*
							FROM
								(
									VALUES
										(
											'1983-11-02':::DATE,
											'\x3187':::BYTES,
											e'dx.7Nh2\x1f6':::STRING,
											NULL,
											2718087910:::OID,
											352468798664229.9458:::DECIMAL,
											'6280:2c13:54d8:36b0:130a:e93d:3ac9:a299/16':::INET
										),
										(
											'1990-03-21':::DATE,
											'\x4ccbec':::BYTES,
											e'->\x06\x010o\\':::STRING,
											389858332:::OID,
											2464749813:::OID,
											2.708296408940698616E+34:::DECIMAL,
											'142.46.85.124/6':::INET
										),
										(
											'1995-08-09':::DATE,
											'\x':::BYTES,
											e'\'':::STRING,
											1495495046:::OID,
											223539051:::OID,
											8.193026533765357783E+29:::DECIMAL,
											'160.70.172.33/4':::INET
										),
										(
											'1970-02-27':::DATE,
											'\x':::BYTES,
											'':::STRING,
											3400475367:::OID,
											2201025256:::OID,
											(-3431716.414289748539):::DECIMAL,
											'a228:61c9:ba3a:68dd:b7b8:2876:9f96:cfa9/55':::INET
										),
										(
											'1988-03-09':::DATE,
											NULL,
											'':::STRING,
											3187816278:::OID,
											619131803:::OID,
											4.006330167783352015E+35:::DECIMAL,
											'180.147.124.106/2':::INET
										),
										(
											'1997-04-10':::DATE,
											'\x41edb86f':::BYTES,
											e'\x17!\x15o4&"':::STRING,
											3284583121:::OID,
											3154207902:::OID,
											(-5.967789328961945427E+19):::DECIMAL,
											'132.125.141.160':::INET
										)
								)
									AS tab_7213
										(col_21907, col_21908, col_21909, col_21910, col_21911, col_21912, col_21913)
						)
				SELECT
					6153146211197169488:::INT8 AS col_21914,
					'23:38:09.725501':::TIME AS col_21915,
					'f5c:f0b9:f13f:24a1:899b:3e3c:d370:6066/70':::INET AS col_21916,
					tab_7215.col3 AS col_21917,
					tab_7215.col4 AS col_21918,
					tab_7215.col1 AS col_21919,
					tab_7215.col6 AS col_21920,
					false AS col_21921,
					tab_7215.col8 AS col_21922,
					tab_7215.col9 AS col_21923,
					tab_7215.col10 AS col_21924,
					tab_7215.col11 AS col_21925
				FROM
					defaultdb.public.table0 AS tab_7215
				WHERE
					tab_7215.col7
				RETURNING
					NULL AS ret_2710, '\xc7':::BYTES AS ret_2711
			],
			defaultdb.public.table0 AS tab_7216
		WHERE
			false
		ORDER BY
			col_21926, col_21926 DESC
		LIMIT
			1:::INT8
	)
		AS col_21927,
	tab_7207.col4 AS col_21928,
	'16:15:48.168747':::TIME AS col_21929,
	1650986085:::OID AS col_21930,
	NULL AS col_21931,
	tab_7207.col2 AS col_21932,
	tab_7207.col11 AS col_21933,
	NULL AS col_21934,
	'j~%"':::STRING AS col_21935
FROM
	defaultdb.public.table0 AS tab_7207
WHERE
	tab_7207.col7
ORDER BY
	tab_7207.col3 DESC, tab_7207.col0 DESC
LIMIT
	74:::INT8;
```

**Comment by rytaft** [comment]
I was able to reproduce it -- the key is to run with RaceEnabled = true. I'll work on fixing this -- thanks for the report!

**Comment by rytaft** [comment]
Here's what seems to have happened:
There was a join expression `((A JOIN B) JOIN C)`. `A` was a zero-cardinality expression containing side effects and mutations, `B` was a zero-cardinality `VALUES` expression, and `C` had an outer column but no side effects. The `AssociateJoin` rule was applied, so that a new expression was created: `(A JOIN (B JOIN C))`. The expression `(B JOIN C)` was now also a zero-cardinality expression, and since it had no side effects, it matched the `SimplifyZeroCardinalityGroup` rule, thus replacing it with a zero-cardinality `VALUES` expression and eliminating the outer column. This is why the outer columns differ between the two otherwise logically-equivalent expressions. 

Basically, the issue seems to be the side effects present in `A`, since otherwise the whole thing would reduce to a zero-cardinality `VALUES` expression. Not sure how we should handle this.

cc @justinj, @andy-kimball

**Comment by andy-kimball** [comment]
@RaduBerinde has a comment on that check that has proven to be prophetic:
```
	// TODO(radu): these checks might be overzealous - conceivably a
	// subexpression with outer columns/side-effects/placeholders could be
	// elided.
	if !r.OuterCols.Equals(other.OuterCols) {
		panic(pgerror.NewAssertionErrorf("outer cols mismatch: %s vs %s", log.Safe(r.OuterCols), log.Safe(other.OuterCols)))
	}
```

**Comment by RaduBerinde** [comment]
Yeah, we can remove those checks. I don't think they caught any bugs anyway. Would be good to add a testcase somewhere like the one described above.

**Comment by rytaft** [comment]
Cool - I just opened a PR to remove those checks and also added a test for the scenario I described above (basically a simplified version of this sqlsmith query).

Since there could be so many differences in logical properties between expressions in the same memo group, should we reconsider only having one copy of the logical props per group? Seems like we could be missing out on some optimizations if we don't have the right logical properties. But then again, maybe it's only relevant for edge cases like this and not worth the extra computation.
